src/people.rs

```diff
@@ -1204,15 +1207,53 @@ impl ContextPeopleExt for Context {
         );
     }
 
-    fn sample_person<R: RngId + 'static>(&self, rng_id: R) -> Result<PersonId, IxaError>
+    #[allow(clippy::cast_sign_loss, clippy::cast_possible_truncation)]
+    fn sample_person<R: RngId + 'static, T: Query>(
+        &self,
+        rng_id: R,
+        q: T,
+    ) -> Result<PersonId, IxaError>
     where
         R::RngType: Rng,
     {
         if self.get_current_population() == 0 {
             return Err(IxaError::IxaError(String::from("Empty population")));
         }
-        let result = self.sample_range(rng_id, 0..self.get_current_population());
-        Ok(PersonId { id: result })
+
+        // Special case the empty query because we can do it in O(1).
+        if q.get_query().is_empty() {
+            let result = self.sample_range(rng_id, 0..self.get_current_population());
+            return Ok(PersonId { id: result });
+        }
+
+        T::setup(self);
+
+        // This function implements "Algorithm R" from KIM-HUNG LI
```

Comment:
@jasonasher would you mind taking a look at this

Reply:
I corrected the algorithm and expanded the test a little bit in PR #127

Reply:
Ugh. Missed the W* in L.4.